libp2p/protocols/connectivity/autonatv2/client.nim

```diff
@@ -0,0 +1,214 @@
+): T =
+  let client = T(dialer: dialer, rng: rng, dialBackTimeout: dialBackTimeout)
+
+  # handler for DialBack messages
+  proc handleStream(
+      conn: Connection, proto: string
+  ) {.async: (raises: [CancelledError]).} =
+    try:
+      let dialBack = DialBack.decode(initProtoBuffer(await conn.readLp(DialBackLpSize))).valueOr:
+        trace "Unable to decode DialBack"
+        return
+      if not await client.handleDialBack(conn, dialBack).withTimeout(
+        client.dialBackTimeout
+      ):
+        trace "Sending DialBackResponse timeout"
+        return
+    except CancelledError as exc:
+      raise exc
+    except LPStreamRemoteClosedError as exc:
+      debug "Connection closed by peer", description = exc.msg, peer = conn.peerId
+    except CatchableError as exc:
+      debug "Exception in handler", description = exc.msg, conn
+
+  client.handler = handleStream
+  client.codec = $AutonatV2Codec.DialBack
+  client
+
+proc handleDialDataRequest*(
+    conn: Connection, req: DialDataRequest
+): Future[DialResponse] {.
+    async: (raises: [CancelledError, AutonatV2Error, LPStreamError])
+.} =
+  debug "Received DialDataRequest",
+    numBytes = req.numBytes, maxAcceptedNumBytes = MaxAcceptedDialDataRequest
+
+  if req.numBytes > MaxAcceptedDialDataRequest:
+    raise newException(
+      AutonatV2Error, "Rejecting DialDataRequest: numBytes is greater than the maximum"
+    )
+
+  # send required data
+  var msg = AutonatV2Msg(
+    msgType: MsgType.DialDataResponse,
+    dialDataResp: DialDataResponse(data: newSeq[byte](MaxDialDataResponsePayload)),
+  )
+  let messagesToSend =
+    (req.numBytes + MaxDialDataResponsePayload - 1) div MaxDialDataResponsePayload
+  for i in 0 ..< messagesToSend:
+    # try:
+    #   await conn.writeLp(msg.encode().buffer)
+    #   debug "Sending DialDataResponse", i = i, messagesToSend = messagesToSend
+    # except CatchableError as exc:
+    #   raise newException(AutonatV2Error, "Failed to send DialDataResponse", exc)
+    await conn.writeLp(msg.encode().buffer)
+    debug "Sending DialDataResponse", i = i, messagesToSend = messagesToSend
+
+  # get DialResponse
+  msg = AutonatV2Msg.decode(initProtoBuffer(await conn.readLp(AutonatV2MsgLpSize))).valueOr:
+    raise newException(AutonatV2Error, "Unable to decode AutonatV2Msg")
+
+  debug "Receive message", msgType = $msg.msgType
+  if msg.msgType != MsgType.DialResponse:
+    raise
+      newException(AutonatV2Error, "Expecting DialResponse, but got " & $msg.msgType)
+
+  return msg.dialResp
+
+proc checkAddrIdx(
+    self: AutonatV2Client, addrIdx: AddrIdx, testAddrs: seq[MultiAddress], nonce: Nonce
+): bool {.raises: [AutonatV2Error].} =
+  debug "checking addrs", addrIdx = addrIdx, testAddrs = testAddrs, nonce = nonce
+  let dialBackAddrs =
+    try:
+      self.expectedNonces[nonce].get()
+    except KeyError:
+      debug "Not expecting this nonce",
+        nonce = nonce, expectedNonces = self.expectedNonces
+      return false
```

Comment:
This should be slightly more efficient as it will not raise a exception

```suggestion
  if not self.expectedNonces.hasKey[nonce]:
    debug "Not expecting this nonce", nonce = nonce, expectedNonces = self.expectedNonces
    return false
    
  let dialBackAddrs =
    try:
      self.expectedNonces[nonce].get()
    except KeyError:
       doAssert false, "checked with hasKey"
```